Feature/refactor commandParser

seongjin-Park210 opened the pull request:
1. Add exceptions to check the validity of data length. 
2. Refactor and add a function for removing blanks.

seongjin-Park210 committed:
refactor: refactoring in commandParser for removing blanks in a string

create a "removeBlanks" function for removing blanks in options / columnNames

seongjin-Park210 committed:
feat: add an exception in commandParser for invalid employee length

feat: add an exception in commandParser for invalid employee length.
The cmd length should be 10 (cmd + 3 options + 6 employee member variables).

seongjin-Park210 committed:
feat: add an exception in CommandParser.getColumnData

A command line should take valid number of data. A command for accessing data takes one column name and data or two column name and corresponding data. This commit checks the validity of a command length.

taemin00 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandParser.java:
```diff
@@ -37,16 +36,20 @@ public static boolean isValidCommand(String cmd) {
 
         for (int i = 0; i < options.size(); i++) {
-            options.set(i, options.get(i).replaceAll("\\p{Z}", ""));
+            options.set(i, removeBlanks(options.get(i)));
```
removeBranks를 메소드 추출하니 훨씬 코드가 깔끔해 보입니다!

kwonhj0116 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandParser.java:
```diff
@@ -60,17 +63,36 @@ public static Employee getEmployee(String cmdLine) {
 
         return columnData;
     }
+
```
공통 부분을 메소드추출 하니 코드가 한 눈에 들어와서 좋습니다!! 고생하셨습니다~!

kwonhj0116 reviewed (approved):
코드 확인했습니다~ 리팩토링 및 테스트코드 추가 고생하셨습니다!

taemin00 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandParser.java:
```diff
@@ -60,17 +63,36 @@ public static Employee getEmployee(String cmdLine) {
-
         String[] cmd = parseCommandLine(cmdLine);
+
+        final int cmdOptionLen = OPT3_POS + 1;
+        final int oneColumnDataLen = cmdOptionLen + 2;
+        final int twoColumnDataLen = cmdOptionLen + 4;
```
length를 사전에 미리 정의해 놓으니 뒤에서 사용하기 좋은 것 같네요!

taemin00 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandParser.java:
```diff
@@ -60,17 +63,36 @@ public static Employee getEmployee(String cmdLine) {
+        final int cmdOptionLen = OPT3_POS + 1;
+        final int oneColumnDataLen = cmdOptionLen + 2;
+        final int twoColumnDataLen = cmdOptionLen + 4;
+        boolean validDataLength = isValidDataLength(cmd, oneColumnDataLen)
+                                    || isValidDataLength(cmd, twoColumnDataLen);
+        if (!validDataLength) {
+            throw new IllegalArgumentException("Invalid Data Length");
+        }
```
isValidDataLength를 오버라이딩 하여서, 인자가 한개일 때와 인자가 2개일 때 처리를 하는 함수 2개를 만들면 어떨까요?
그렇게 되면 isValidDataLength()를 2번 호출 할 필요가 없게 되고,
validDataLength 라는 로컬 변수를 쓸 필요가 없게 될 것 같습니다!

mj610-kim commented on EmployeeManagementSystem/src/test/java/CommandParserTest.java:
```diff
@@ -12,7 +12,10 @@
+    String invalidCmdLine = "MOE,-p,-m, ,birthday,09,cl,CL2";
+    String invalidEmpDataLength = "ADD, , , ,18050301,KYUMOK KIM,CL2,010-9777-6055";
+    String invalidTwoColDataLen = "MOD,-p,, ,name,KYUMOK KIM,name";
+    String invalidOneColDataLen = "DEL, , , ,birthday";
```
테스트코드에도 변수명에 자세한 설명이 있다보니 해석이 더욱 쉬워진 것 같습니다. 감사합니다!

mj610-kim commented on EmployeeManagementSystem/src/test/java/CommandParserTest.java:
```diff
@@ -23,13 +26,14 @@ public void getCommandTest() {
     public void getEmployeeTest() {
         String birthday = cmdParser.getEmployee(cmdLine1).getBirthDayAll();
         assertEquals("19980906", birthday);
+        assertThrows(IllegalArgumentException.class, () -> cmdParser.getEmployee(invalidEmpDataLength));
```
추가적인 검증 감사합니다!

taemin00 commented on EmployeeManagementSystem/src/main/java/com/cra08/excellentcode/CommandParser.java:
```diff
@@ -60,17 +63,36 @@ public static Employee getEmployee(String cmdLine) {
+    }
+
+
+    public static boolean isValidDataLength(String[] cmd, int validLength) {
+        return cmd.length == validLength;
+    }
```
이 부분을 오버라이딩 하여 isValidDataLength(String[] cmd, int validLength1, int validLength2)라는 함수를 하나 더 만들면 어떨까하네요 ㅎㅎ

seongjin-Park210 replied:
넵! 의견 주신대로 오버로딩 함수 추가했습니다~ 좋은 의견 감사합니다.

mj610-kim reviewed (approved):
가독성이 더 좋아진 것 같습니다. 고생하셨습니다!

seongjin-Park210 committed:
refactor: overload isValidDataLength

Overload the function isValidDataLength so that it can accept two validLengths.
